src/ess/amor/conversions.py

```diff
@@ -1,24 +1,168 @@
-        source_chopper_1_position + source_chopper_2_position
-    ) * sc.scalar(0.5)
-    return sample_position - chopper_midpoint
+    return (
+        theta
+        - sample_rotation
+        #
+        - angle_to_center_of_beam
+        # "kappa" the natural incidence angle
+        # is a fixed parameter of the instrument
+        - sc.scalar(0.245, unit='deg').to(unit='rad')
+    )
+
+
+def wavelength(
+    event_time_offset, divergence_angle, L1, L2, chopper_phase, chopper_frequency
+):
+    "Converts event_time_offset to wavelength using the chopper settings."
+    out = event_time_offset.to(unit="ns", dtype="float64", copy=True)
+    unit = out.bins.unit
+    tau = (1 / (2 * chopper_frequency)).to(unit=unit)
+    tof_offset = tau * chopper_phase / (180.0 * sc.units.deg)
+
+    minimum = -tof_offset
+    frame_bound = tau - tof_offset
+    maximum = 2 * tau - tof_offset
+
+    # Frame unwrapping
+    out += sc.where(
+        (minimum < event_time_offset) & (event_time_offset < frame_bound),
+        tof_offset,
+        sc.where(
+            (frame_bound < event_time_offset) & (event_time_offset < maximum),
+            tof_offset - tau,
+            sc.scalar(np.nan, unit=unit),
+        ),
+    )
+    # Correction for path length through guides being different
+    # depending on incident angle.
+    out -= (divergence_angle.to(unit="deg") / (180.0 * sc.units.deg)) * tau
+    out *= (sc.constants.h / sc.constants.m_n) / (L1 + L2)
+    return out.to(unit='angstrom', copy=False)
+
+
+def _not_between(v, a, b):
+    return (v < a) | (v > b)
+
+
+def add_common_coords_and_masks(
+    da: RawDetectorData[RunType],
+    ylim: YIndexLimits,
+    zlims: ZIndexLimits,
+    bdlim: BeamDivergenceLimits,
+    wbins: WavelengthBins,
+    beam_size: BeamSize[RunType],
+) -> ReducibleData[RunType]:
+    "Adds coords and masks that are useful for both reference and sample measurements."
+    da = da.transform_coords(
+        ("wavelength", "theta", "angle_of_divergence", "Q"),
+        {
+            "divergence_angle": "pixel_divergence_angle",
+            "wavelength": wavelength,
+            "theta": theta,
+            "angle_of_divergence": angle_of_divergence,
+            "Q": reflectometry_q,
+        },
+        rename_dims=False,
+        keep_intermediate=False,
+    )
+    da.masks["stripe_range"] = _not_between(da.coords["stripe"], *ylim)
+    da.masks['z_range'] = _not_between(da.coords["z_index"], *zlims)
+    da.bins.masks["divergence_too_large"] = _not_between(
+        da.bins.coords["angle_of_divergence"],
+        bdlim[0].to(unit=da.bins.coords["angle_of_divergence"].bins.unit),
+        bdlim[1].to(unit=da.bins.coords["angle_of_divergence"].bins.unit),
+    )
+    da.bins.masks['wavelength'] = _not_between(
+        da.bins.coords['wavelength'],
+        wbins[0],
+        wbins[-1],
+    )
+    da /= footprint_on_sample(
+        da.bins.coords['theta'],
+        beam_size,
+        da.coords['sample_size'],
+    )
```

Comment:
Looking at it from the perspective of regular functions, not Sciline providers, the two operations should be split. They are conceptually different and doing both in one function violates the single responsibility principle.

Whether a provider should call both functions together is a separate issue. But I still think that it should be split. Note that the output currently is also `FootprintCorrectedData`. But you could insert an intermediate step for `MaskedData`. This would allow users to customise the masks. As it stand now, normalisation and masking are an atomic operation that can only be customised as a whole. Plus, it hides the fact that those operations happen in the graph.

Imagine a user looking at the graph. They may want to know where masks are applied. How would they find out? Or the other way around, how would they find out where footprint corrections are applied?

Reply:
This is now split into separate functions. I haven't added the extra steps in the graph because as I mentioned I don't think they make it better.

> Imagine a user looking at the graph. They may want to know where masks are applied. How would they find out? Or the other way around, how would they find out where footprint corrections are applied?

We talked about this so you already know, but for anyone else reading this:
I doubt it wouldn't be difficult for them to find out. There is only one step between `RawDetectorData` and `ReducibleData` and that is where they find it (provider called `add_coords_masks_and_apply_corrections`). They could also grep the source code or search the documentation.

Reply:
> This would allow users to customise the masks.

Customizing masks is already possible. Additional masks can be added to `ReducibleData` just as well as they could have be added to `MaskedData`.